### PR TITLE
Improved bond highlights

### PR DESCRIPTION
**Pull request by DavidACosgrove**
#### Reference Issue
No Issue

#### What does this implement/fix? Explain your changes.
This PR improves the way bond highlights are drawn.
1. If using the default continuousHighlight, it mitres the lines so that they meet nicely:
image
vs
image
2.  Tidied up the bond highlights when not doing continuousHighlight or circleAtoms:
image
vs
image
The changes in the latter case are more subtle - in this example, because the N is only connected by highlighted bonds, the N gets the highlight colour even though it isn't an explicitly highlighted atom.  The ether O isn't in highlight colour because one of its bonds isn't highlighted.  This makes the atom symbol colour scheme more consistent with the carbon skeleton colouring.  In both the new and old code, atom 8 is highlighted in green, but bond 8 isn't, so half of bond 8 is green.
#### Any other comments?
After 1 too many tedious sessions of updated lots of hash codes, I've added a script in $RDBASE/Code/GraphMol/MolDraw2D that does it automatically based on the output of the relevant test program.

### CODE REVIEW

**Comment by greglandrum** [comment]
@DavidACosgrove the test failures here are real

**Comment by DavidACosgrove** [comment]
Sorry. I will put on a note on my monitor to not forget to check the python unit tests before launching the PR.

**Comment by greglandrum** [comment]
> Sorry. I will put on a note on my monitor to not forget to check the python unit tests before launching the PR.

Eh, no worries, that's what the automated tests are for! 😀

File: Code/GraphMol/MolDraw2D/Wrap/testMolDraw2D.py [hunk 1]
```diff
@@ -185,8 +185,12 @@ def testDrawMoleculesArgs(self):
                          highlightAtomColors=acolors)
     drawer.FinishDrawing()
     svg = drawer.GetDrawingText()
+    with open('testDrawMoleculeArgs_1.svg', 'w') as f:
+      f.write(svg)
     # 4 molecules, 6 bonds each:
-    self.assertEqual(svg.count('fill:none;fill-rule:evenodd;stroke:#FF7F7F'), 24)
+    re_str = r"path class='bond-\d+ atom-\d+ atom-\d+' d='M \d+.\d+,\d+.\d+ L \d+.\d+,\d+.\d+ L \d+.\d+,\d+.\d+ L \d+.\d+,\d+.\d+ Z' style='fill:#FF7F7F;"
```
Comment (greglandrum) [inline Code/GraphMol/MolDraw2D/Wrap/testMolDraw2D.py:189]: oh gracious... another regex.
:nauseated_face:

**Review by greglandrum — CHANGES_REQUESTED** [review]
There are a lot of code changes here relative to what's documented in the PR description and without some more documentation, at least in the PR itself, it's really hard to tell why they are there.

Also: the new code is not a good stylistic match to the rest of the code. There are a bunch of explicit variable types that could be autos, multiple variables being declared on the same line, etc.

File: Code/GraphMol/MolDraw2D/Wrap/testMolDraw2D.py (hunk 1 above)
Comment (DavidACosgrove) [reply Code/GraphMol/MolDraw2D/Wrap/testMolDraw2D.py:189]: I am getting quite attached to them for test purposes.

File: Code/GraphMol/MolDraw2D/DrawMol.cpp [hunk 2]
```diff
@@ -1892,11 +1891,27 @@ std::pair<DrawColour, DrawColour> DrawMol::getBondColours(Bond *bond) {
...
+      col1 = getColour(at1_idx);
+      Atom *at1 = drawMol_->getAtomWithIdx(at1_idx);
+      const Bond *bond1 = atomHighlightedBond(at1);
+      if (bond1 != nullptr) {
+        DrawColour bondCol = getHighlightBondColour(
+            bond1->getIdx(), drawOptions_, highlightBonds_, highlightBondMap_);
+        if (col1 == bondCol) {
+          col1 = getColourByAtomicNum(at1->getAtomicNum(), drawOptions_);
+        }
+      }
```
[the remark this answers is not in the document]
Comment (DavidACosgrove) [reply Code/GraphMol/MolDraw2D/DrawMol.cpp:1903]: Yep - 4 weeks out from when I wrote it, I am struggling too.

[the remark this answers is not in the document]
Comment (DavidACosgrove) [reply Code/GraphMol/MolDraw2D/DrawMol.cpp:1903]: Your instincts are clearly well honed after your time off.  None of the regression tests were using it.  I assume it was left over from a previous attempt at tackling the bond colouring.  I've removed it and all tests still passed.  Sorry.

**Review by greglandrum — COMMENTED** [review]
LGTM

**Review by greglandrum — APPROVED** [review]

**Pull request merged**